app/models/repository_host/repository_host_data.rb

```diff
@@ -0,0 +1,89 @@
+# frozen_string_literal: true
+
+# This class is meant to be a facade over the raw upstream data coming
+# from the different repository hosts' repository data. It's main goal
+# is to standardize the output from each repository host into a concrete
+# set of data so we can make sure the raw data is being mapped to the same
+# schema within the Libraries.io models and code.
+class RepositoryHost::RepositoryHostData
+  attr_reader :archived, :default_branch, :description, :fork, :full_name, :has_issues, :has_wiki, :homepage, :host_type,
+              :keywords, :language, :license, :name, :owner, :parent, :is_private, :repository_uuid, :scm, :size
+
+  def initialize(
+    archived:,
+    default_branch:,
+    description:,
+    fork:,
+    full_name:,
+    has_issues:,
+    has_wiki:,
+    homepage:,
+    host_type:,
+    keywords:,
+    language:,
+    license:,
+    name:,
+    owner:,
+    parent:,
+    is_private:,
+    repository_uuid:,
+    scm:,
+    size:,
+    **kwargs
+  )
+    @archived = archived
+    @default_branch = default_branch
+    @description = description
+    @fork = fork
+    @full_name = full_name
+    @has_issues = has_issues
+    @has_wiki = has_wiki
+    @homepage = homepage
+    @host_type = host_type
+    @keywords = keywords
+    @language = language
+    @license = license
+    @name = name
+    @owner = owner
+    @parent = parent
+    @private = is_private
+    @repository_uuid = repository_uuid
+    @scm = scm
+    @size = size
+
+    raise "Unexpected arguments sent: #{kwargs.keys.join(', ')}" unless kwargs.keys.empty?
+  end
```

Comment:
```suggestion
RepositoryHost::RepositoryHostData = Struct.new(
  :archived, :default_branch, :description, :fork, :full_name, :has_issues, :has_wiki, :homepage, :host_type,
  :keywords, :language, :license, :name, :owner, :parent, :is_private, :repository_uuid, :scm, :size,
  keyword_init: true
) do
```

This could be DRYed up by making `RepositoryHostData` a `Struct` with `keyword_init: true`.

That approach would mean you couldn't use the `**kwargs` parameter, but I'm not sure you need it. Could the factories pass the keyword arguments directly instead of creating a hash that is later splatted out? This also has the benefit of making key mismatch errors for callers catchable via static analysis, whereas this approach defers that error to runtime.

Reply:
I wasn't sure what direction this class was going to end up going but I think a `Struct` makes sense here, actually a `Data` would be better but we are stuck on an older Ruby version for now. The splat was me being lazy and copying code around, but I can switch it over to keywords directly now that I'm more confident on what keys are expected to exist. I'll make these changes today.

Reply:
🎱 I'm wondering if it makes sense to have the callers just pass in the raw data hash, and then have this object (or maybe a controller/interaction) figure out how to convert all the inputs into the appropriate data models to be persisted, vs having a bunch of factories/temp objects floating around. Mostly thinking that the factory classes aren't really doing a ton, or represent objects that will be reused a lot. 

I'm thinking something like: 

```
RawUpstreamDataToDatabaseModelConverter

def convert_gitlab_repository_raw_data(data_hash)
  repo_hash = {
      repository_uuid: data_hash.api_project.id,
      host_type: "GitLab",
      fork: data_hash.api_project.try(:forked_from_project).present?,
      scm: "git",
      is_private: api_project.visibility != "public",
    }

   create_the_thing(repo_hash)
end

def convert_github_repository_raw_data(data_hash);

def create_the_thing(data_hash_to_be_persisted)
      Repository.create!(repo_hash) # replace the Repository.create_from_hash
end

end

# some caller that gets the api data
github_api_data = Curl.get("https://www.github.com/foo/bar")

foo = RawUpstreamDataToDatabaseModelConverter.convert_gitlab_repository_raw_data(github_api_data) 
# you can play around with how much of that data should go into some sort of initializer in the converter, the caller, or the converter method

return foo
```

I think this has the advantage of keeping all the conversion logic in one place (easier to DRY up), no need to have intermediary temporary/facade objects, and a clean & consistent interface/pattern for the upstream to use (e.g. with error handling on bad input data). The single responsibility here could be something like "takes in garbage upstream stuff and turns it into our own objects."

Reply:
I like the single-factory idea from @jhan217 , additionally because it makes it easy to see the differences among all the sources in one place too.